src/profile-logic/comparison.js

```diff
@@ -246,3 +309,498 @@ function adjustNullableCategories(
+    const translationMap = new Map();
+    const funcTranslationMap = translationMapsForFuncs[threadIndex];
+    const categoryTranslationMap = translationMapsForCategories[threadIndex];
+
+    for (let i = 0; i < frameTable.length; i++) {
+      const addressIndex = frameTable.address[i];
+      const address =
+        typeof addressIndex === 'number' && addressIndex >= 0
+          ? stringTable.getString(addressIndex)
+          : null;
+      const category = frameTable.category[i];
+      const newCategory =
+        category === null ? null : categoryTranslationMap.get(category);
+      if (newCategory === undefined) {
+        throw new Error(stripIndent`
+          We couldn't find the category of frame ${i} in the translation map.
+          This is a programming error.
+        `);
+      }
+
+      const newFunc = funcTranslationMap.get(frameTable.func[i]);
+      if (newFunc === undefined) {
+        throw new Error(stripIndent`
+          We couldn't find the function of frame ${i} in the translation map.
+          This is a programming error.
+        `);
+      }
+
+      const implementationIndex = frameTable.implementation[i];
+      const implementation =
+        typeof implementationIndex === 'number'
+          ? stringTable.getString(implementationIndex)
+          : null;
+
+      newFrameTable.address.push(
+        address === null ? -1 : newStringTable.indexForString(address)
+      );
+      newFrameTable.category.push(newCategory);
+      newFrameTable.func.push(newFunc);
+      newFrameTable.implementation.push(
+        implementation === null
+          ? null
+          : newStringTable.indexForString(implementation)
+      );
+      newFrameTable.line.push(frameTable.line[i]);
+      newFrameTable.column.push(frameTable.column[i]);
+      newFrameTable.optimizations.push(frameTable.optimizations[i]);
+
+      translationMap.set(i, newFrameTable.length);
+      newFrameTable.length++;
+    }
+
+    translationMaps.push(translationMap);
+  });
+
+  return { frameTable: newFrameTable, translationMaps };
+}
+
+/**
+ * This combines the stack tables for a list of threads. It returns the new
+ * stack table with the translation maps to be used in subsequent merging
+ * functions.
+ */
+function combineStackTables(
+  translationMapsForCategories: TranslationMapForCategories[],
+  translationMapsForFrames: TranslationMapForFrames[],
+  ...threads: Thread[]
+): { stackTable: StackTable, translationMaps: TranslationMapForStacks[] } {
+  const translationMaps = [];
+  const newStackTable = getEmptyStackTable();
+
+  threads.forEach((thread, threadIndex) => {
+    const { stackTable } = thread;
+    const translationMap = new Map();
+    const frameTranslationMap = translationMapsForFrames[threadIndex];
+    const categoryTranslationMap = translationMapsForCategories[threadIndex];
+
+    for (let i = 0; i < stackTable.length; i++) {
+      const newFrameIndex = frameTranslationMap.get(stackTable.frame[i]);
```

Comment:
Optional: I believe these types of checks can use the `ensureExists` function.

Reply:
Oh hum, unless they are null values.

Reply:
Hummm I think we can't get null values from the translation maps. So maybe I can use `ensureExists` indeed.
I'll try it and see if this looks more readable. (probably it does)

Reply:
So as we discussed, [I tried](https://github.com/firefox-devtools/profiler/pull/1883/commits/9b931b337446624c7c249b42cf977d4e13a634f8) and this isn't really better, so I'll remove this commit from the commit list.